feat(android): GLCanvasElement

ramezgerges opened the pull request:
GitHub Issue (If applicable): https://github.com/unoplatform/uno/issues/9405, follows from #16621

<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal), unless the change is documentation related. -->

## PR Type

What kind of change does this PR introduce?
<!-- Please uncomment one or more that apply to this PR

- Bugfix
- Feature
- Code style update (formatting)
- Refactoring (no functional changes, no api changes)
- Build or CI related changes
- Documentation content changes
- Project automation
- Other... Please describe:

-->

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->


## What is the new behavior?

<!-- Please describe the new behavior after your modifications. -->

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

## Other information

<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

ramezgerges committed:
chore: make Graohics3DGL build on Android

ramezgerges committed:
chore: more Graohics3DGL adjustments for Android

ramezgerges committed:
chore: use Silk.NET.OpenGL even on GLES platforms

ramezgerges committed:
feat: GLCanvasElement on Android

ramezgerges committed:
fix(writablebitmap): invalidating a WritableBitmap on Android is not updating

ramezgerges committed:
chore: GLCanvasElement_Cube sample on Android

ramezgerges committed:
chore: dynamically compute GLSL version

unodevops commented:
🤖 Your WebAssembly Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-18216/index.html

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-18216/index.html

ramezgerges committed:
chore: minor doc changes

ramezgerges committed:
chore: packaging

jeromelaban commented on src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.Android.cs:
```diff
@@ -43,11 +43,6 @@ partial void UpdatePlatform()
 			var newState = new BorderLayerState(drawArea.Size, _borderInfoProvider);
 			var previousLayoutState = _currentState;
 
-			if (newState.Equals(previousLayoutState))
-			{
-				return;
-			}
```
This looks like a performance issue. Do you recall why you needed to remove it?

ramezgerges replied:
I managed to create a repro that's completely independent of GLCanvasElement to illustrate the problem and opened a separate PR.
https://github.com/unoplatform/uno/pull/18408

jeromelaban commented on src/Uno.UI.RuntimeTests/Uno.UI.RuntimeTests.Skia.csproj:
```diff
@@ -60,10 +60,13 @@
 	</ItemGroup>
 
 	<ItemGroup>
+<!--		<ProjectReference Include="..\Uno.Foundation\Uno.Foundation.Reference.csproj" />-->
+<!--		<ProjectReference Include="..\Uno.UWP\Uno.Reference.csproj" />-->
```
Is this still needed?

Youssef1313 committed:
chore: Fix race

Building Uno.UI.Reference and Uno.UI.Skia in parallel was writing to the same mergedstyles.xaml, causing random crashes at runtime

ramezgerges committed:
chore: more packaging

ramezgerges committed:
chore: import targetframework-override.props in Graphics3DGL.csproj and remove NetPrevious-android from TargetFrameworks

jeromelaban commented:
/azp run

unodevops commented:
🤖 Your WebAssembly Sample App stage site is ready! Visit it here: https://unowasmprstaging.z20.web.core.windows.net/pr-18216/index.html

unodevops commented:
🤖 Your Docs stage site is ready! Visit it here: https://unodocsprstaging.z13.web.core.windows.net/pr-18216/index.html

jeromelaban commented:
@ramezgerges 

```
Section changed: SearchContent
fail: Microsoft.UI.Xaml.UIElement[0]
      OnElementUnloaded failed in FrameworkElement
      System.NullReferenceException: Object reference not set to an instance of an object.
         at UITests.Shared.Windows_UI_Composition.RotatingCubeGlCanvasElement.OnDestroy(GL Gl) in C:\a\1\s\src\SamplesApp\UITests.Shared\Windows_UI_Composition\RotatingCubeGlCanvasElement.cs:line 121
         at Uno.WinUI.Graphics3DGL.GLCanvasElement.OnUnloaded(Object sender, RoutedEventArgs routedEventArgs) in C:\a\1\s\src\AddIns\Uno.WinUI.Graphics3DGL\GLCanvasElement.cs:line 286
         at Microsoft.UI.Xaml.FrameworkElement.<OnFwEltUnloaded>g__InvokeUnloaded|86_0() in C:\a\1\s\src\Uno.UI\UI\Xaml\FrameworkElement.crossruntime.cs:line 206
         at Microsoft.UI.Xaml.FrameworkElement.<OnFwEltUnloaded>g__InvokeUnloadedWithTry|86_1() in C:\a\1\s\src\Uno.UI\UI\Xaml\FrameworkElement.crossruntime.cs:line 221
info: Uno.UI.Runtime.Skia.MacOS.MacOSNativeOpenGLWrapper[0]
      Found EGL version 1.5.
fail: Microsoft.UI.Xaml.UIElement[0]
      OnElementLoaded failed in FrameworkElement
      System.Exception: Vertex shader failed to compile: ERROR: 0:2: '#' : invalid character
      ERROR: 0:2: 'for' : syntax error

         at UITests.Shared.Windows_UI_Composition.SimpleTriangleGlCanvasElement.Init(GL gl) in C:\a\1\s\src\SamplesApp\UITests.Shared\Windows_UI_Composition\SimpleTriangleGlCanvasElement.cs:line 80
         at Uno.WinUI.Graphics3DGL.GLCanvasElement.OnLoaded(Object sender, RoutedEventArgs routedEventArgs) in C:\a\1\s\src\AddIns\Uno.WinUI.Graphics3DGL\GLCanvasElement.cs:line 245
         at Microsoft.UI.Xaml.FrameworkElement.<OnFwEltLoaded>g__InvokeLoadedWithTry|327_1() in C:\a\1\s\src\Uno.UI\UI\Xaml\FrameworkElement.Layout.crossruntime.cs:line 111
dbug: Uno.UI.Runtime.Skia.MacOS.MacOSWindowHost[0]
      SetPointerCapture not supported on macOS.
dbug: Uno.UI.Runtime.Skia.MacOS.MacOSWindowHost[0]
      ReleasePointerCapture not supported on macOS.
Section changed: SearchContent
info: Uno.UI.Runtime.Skia.MacOS.MacOSNativeOpenGLWrapper[0]
      Found EGL version 1.5.
fail: Microsoft.UI.Xaml.UIElement[0]
      OnElementLoaded failed in FrameworkElement
      System.Exception: Error compiling shader of type VertexShader, failed with error ERROR: 0:2: '#' : invalid character
      ERROR: 0:2: 'for' : syntax error

         at UITests.Shared.Windows_UI_Composition.RotatingCubeGlCanvasElement.Shader.LoadShader(ShaderType type, String src) in C:\a\1\s\src\SamplesApp\UITests.Shared\Windows_UI_Composition\RotatingCubeGlCanvasElement.cs:line 241
         at UITests.Shared.Windows_UI_Composition.RotatingCubeGlCanvasElement.Shader..ctor(GL gl, String vertexShaderSource, String fragmentShaderSource) in C:\a\1\s\src\SamplesApp\UITests.Shared\Windows_UI_Composition\RotatingCubeGlCanvasElement.cs:line 175
         at UITests.Shared.Windows_UI_Composition.RotatingCubeGlCanvasElement.Init(GL Gl) in C:\a\1\s\src\SamplesApp\UITests.Shared\Windows_UI_Composition\RotatingCubeGlCanvasElement.cs:line 113
         at Uno.WinUI.Graphics3DGL.GLCanvasElement.OnLoaded(Object sender, RoutedEventArgs routedEventArgs) in C:\a\1\s\src\AddIns\Uno.WinUI.Graphics3DGL\GLCanvasElement.cs:line 245
         at Microsoft.UI.Xaml.FrameworkElement.<OnFwEltLoaded>g__InvokeLoadedWithTry|327_1() in C:\a\1\s\src\Uno.UI\UI\Xaml\FrameworkElement.Layout.crossruntime.cs:line 111

```

MartinZikmund commented on src/Uno.UI-Skia-only.slnf:
```diff
@@ -54,6 +54,7 @@
       "Uno.UI.Toolkit\\Uno.UI.Toolkit.Skia.csproj",
       "Uno.UI.XamlHost.Skia.Wpf\\Uno.UI.XamlHost.Skia.Wpf.csproj",
       "Uno.UI.XamlHost\\Uno.UI.XamlHost.Skia.csproj",
+      "Uno.UI\\Uno.UI.Reference.csproj",
```
Please also add `Uno.Reference` and `Uno.Foundation.Reference` otherwise browsing code in `Uno.UI` makes VS always complain about missing types when the files are viewed without explicitly selecting Skia target

jeromelaban commented:
We won't need this change, this is specifically for native.
